[RDY] vim-patch:8.0.{1037,1046,1361}

janlazo opened the pull request:
**vim-patch:8.0.1037: "icase" of 'diffopt' is not used for highlighting**

Problem:    "icase" of 'diffopt' is not used for highlighting differences.
Solution:   Also use "icase". (Rick Howe)
https://github.com/vim/vim/commit/da22b8cc8b1b96fabd5a4c35c57b04a351340fb1

**vim-patch:8.0.1046: code duplication in diff mode**

Problem:    Code duplication in diff mode.
Solution:   Use diff_equal_char() also in diff_cmp(). (Rick Howe)
vim/vim@ae96b8d

**vim-patch:8.0.1361: some users don't want to diff with hidden buffers**

Problem:    Some users don't want to diff with hidden buffers.
Solution:   Add the "hiddenoff" item to 'diffopt'. (Alisue, closes vim/vim#2394)
vim/vim@97ce419

janlazo committed:
vim-patch:8.0.1037: "icase" of 'diffopt' is not used for highlighting

Problem:    "icase" of 'diffopt' is not used for highlighting differences.
Solution:   Also use "icase". (Rick Howe)
https://github.com/vim/vim/commit/da22b8cc8b1b96fabd5a4c35c57b04a351340fb1

janlazo committed:
vim-patch:8.0.1046: code duplication in diff mode

Problem:    Code duplication in diff mode.
Solution:   Use diff_equal_char() also in diff_cmp(). (Rick Howe)
https://github.com/vim/vim/commit/ae96b8d058cffd9d07b78cb7a9ccd382185b9dd6

ZviRackover commented on src/nvim/diff.c:
```diff
@@ -1870,6 +1870,35 @@ bool diffopt_horizontal(void)
+  }
+  if (l > 1) {
+    if (STRNCMP(p1, p2, l) != 0
+        && (!enc_utf8
```
You can drop use of `enc_utf8` as there is an effort to deprecate it.

janlazo replied:
I'll do it in a separate commit so it's easier to review.

janlazo committed:
diff: drop enc_utf8 check in diff_equal_char()

enc_utf8 is always true for nvim.

justinmk commented:
```
Found errors in Test_diff_hlID():
function RunTheTest[35]..Test_diff_hlID line 11: Expected 'DiffAdd' but got 'DiffChange'
function RunTheTest[35]..Test_diff_hlID line 12: Expected 'DiffDelete' but got 'DiffText'
function RunTheTest[35]..Test_diff_hlID line 14: Expected 'FoldColumn' but got 'DiffAdd'
function RunTheTest[35]..Test_diff_hlID line 18: Expected 'DiffAdd' but got 'DiffChange'
```

janlazo commented:
I'd rather omit 8.0.1153 if it needs 8.0.1123 for the highlight changes because of the toolbar feature.

janlazo commented on src/nvim/diff.c:
```diff
@@ -1597,6 +1597,34 @@ static bool diff_equal_entry(diff_T *dp, int idx1, int idx2)
+// Compare the characters at "p1" and "p2".  If they are equal (possibly
+// ignoring case) return true and set "len" to the number of bytes.
+static bool diff_equal_char(const char_u *const p1, const char_u *const p2,
+                            int *const len)
```
I know about the objections to `const` pointers but these params should all be `const` pointers to begin with because of the function and param descriptions. It's a new function so nothing should break.

janlazo committed:
vim-patch:8.0.1361: some users don't want to diff with hidden buffers

Problem:    Some users don't want to diff with hidden buffers.
Solution:   Add the "hiddenoff" item to 'diffopt'. (Alisue, closes vim/vim#2394)
https://github.com/vim/vim/commit/97ce419201421f65f4764549ed80307a7ef9c7a6

janlazo committed:
lint

justinmk commented:
https://github.com/neovim/neovim/commit/47f9ab46357e7f0cbd0c7f475ade01c2c26aac47 should fix the travis macOS problem.
